### PR TITLE
Update UMAP notebook, test precomputed_knn, write data to parquet

### PR DESCRIPTION
**Pull request by rishic3**
- Updated UMAP notebook run with Pyspark+Jupyter, aligning with instructions.
- Added test/functionality for precomputed_knn.
- Updated model writer/reader to save array attributes to parquet via Spark DF.
- Lowered error thresholds to track perf regressions based on empirical tests (bumped to +3 stdevs above mean error)

### CODE REVIEW

**Comment by rishic3** [comment]
build

**Comment by rishic3** [comment]
build

File: python/src/spark_rapids_ml/umap.py [hunk 1]
```diff
@@ -1441,19 +1422,48 @@ def _out_schema(self, input_schema: StructType) -> Union[StructType, str]:
...
+        def write_sparse_array(array: scipy.sparse.spmatrix, df_path: str) -> None:
+            schema = StructType(
+                [
+                    StructField("indices", ArrayType(IntegerType(), False), False),
+                    StructField("indptr", ArrayType(IntegerType(), False), False),
+                    StructField("data", ArrayType(FloatType(), False), False),
+                    StructField("shape", ArrayType(IntegerType(), False), False),
+                ]
+            )
+            data_df = spark.createDataFrame(
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/umap.py:1450]: By putting all data in a single row, what are the limits on data sizes?  Here and below for dense?   Should add these limits to documentation.

Alternative is to create regular possibly multiple dataframes, save in multiple files/subfolders and add a column to order/sort by to restore original order.

File: python/tests/test_umap.py [hunk 2]
```diff
@@ -672,4 +675,97 @@ def test_umap_sparse_vector(
...
+        feature_cols = [f"c{i}" for i in range(X.shape[1])]
+        schema = [f"{c} {pyspark_type}" for c in feature_cols]
+        df = spark.createDataFrame(X.tolist(), ",".join(schema))
+        df = df.withColumn("features", array(*feature_cols)).drop(*feature_cols)
+
+        umap = UMAP(
+            num_workers=gpu_number,
+            metric="sqeuclidean",
+            random_state=random_state,
+            precomputed_knn=precomputed_knn,
```
Comment (eordentlich) [inline python/tests/test_umap.py:738]: Can we convert back to numpy to emphasize that numpy ok here in keeping with our no need for gpu on driver?

**Comment by rishic3** [comment]
build

File: python/src/spark_rapids_ml/umap.py [hunk 3]
```diff
@@ -1441,19 +1423,39 @@ def _out_schema(self, input_schema: StructType) -> Union[StructType, str]:
...
     """
 
     def saveImpl(self, path: str) -> None:
+
+        spark = _get_spark_session()
+
+        def write_sparse_array(array: scipy.sparse.spmatrix, df_dir: str) -> None:
+            indptr_df = spark.createDataFrame(array.indptr, schema=["indptr"])
+            indices_data_df = spark.createDataFrame(
+                pd.DataFrame({"indices": array.indices, "data": array.data})
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/umap.py:1445]: Better to add the row_id here in the pd dataframe in case it is shuffled when creating spark dataframe.

**Comment by rishic3** [comment]
build

**Review by eordentlich — APPROVED** [review]
👍

**Pull request merged**